### PR TITLE
Fix parsing of math.MinInt64.

### LINKED ISSUE #2302: sql: unable to parse math.MinInt64
We currently use `strconv.ParseInt` to parse unsigned integers. Unfortunately, this means we can't parse `-9223372036854775808` because `9223372036854775808` is outside of the range of signed integers. The expression `-9223372036854775808` currently scanned as two tokens, a `-` (unary minus) and an integer. It isn't feasible to scan this as a single token because the scanner doesn't know whether the `-` is a unary minus or a binary minus (e.g. `a-b` or `-b`).

Not sure what the correct solution is here. We could possibly scan `9223372036854775808`as an unsigned int but then store it into the signed int field passed around everywhere. This might get tricky and bug prone in that we have to recognize when the value is being negated that it is already negative. We could add a new `UintVal` expression type and properly deal with the places this is converted into a signed `IntVal`. I think this approach is more likely to be correct, but possibly touches a large swath of code.

### PR DESCRIPTION
**Pull request by petermattis**
Parse integer tokens using strconv.ParseUint, but only allow values in
the range [0, 9223372036854775808]. The resulting value is then cast to
an int64 resulting in values in the range [0, 9223372036854775807] +
-9223372036854775808. During expression normalization, if we see unary
minus of the value -9223372036854775808 we know that we really parsed
9223372036854775808 and are now negating it. During expression
evaluation, if we see a negative IntVal we know that we're attempting to
use (positive) 9223372036854775808 which is outside of the valid int64
range.

Fixes #2302.

### CODE REVIEW

File: sql/parser/eval.go [hunk 1]
```diff
@@ -625,6 +625,9 @@ func EvalExpr(expr Expr) (Datum, error) {
 		return DString(t), nil
 
 	case IntVal:
+		if t < 0 {
+			return DNull, fmt.Errorf("integer value out of range: %d", uint64(t))
```
Comment (tamird) [inline sql/parser/eval.go]: why the cast here?

File: sql/parser/expr.go [hunk 2]
```diff
@@ -211,7 +211,7 @@ func (node BytesVal) String() string {
 type IntVal int64
 
 func (node IntVal) String() string {
-	return strconv.FormatInt(int64(node), 10)
+	return strconv.FormatUint(uint64(node), 10)
```
Comment (tamird) [inline sql/parser/expr.go]: why the cast here?

File: sql/parser/scan.go [hunk 3]
```diff
@@ -535,16 +535,21 @@ func (s *scanner) scanNumber(lval *sqlSymType, ch int) {
...
-		lval.ival, err = strconv.ParseInt(lval.str[2:], 16, 64)
+		uval, err = strconv.ParseUint(lval.str[2:], 16, 64)
 	} else {
-		lval.ival, err = strconv.ParseInt(lval.str, 10, 64)
+		uval, err = strconv.ParseUint(lval.str, 10, 64)
 	}
+	if err == nil && uval > 1<<63 {
+		err = fmt.Errorf("integer value out of range: %d", uval)
+	}
+	lval.ival = int64(uval)
```
Comment (tamird) [inline sql/parser/scan.go]: can you add a comment about overflow here and below?

File: sql/parser/scan_test.go [hunk 4]
```diff
@@ -395,8 +395,8 @@ func TestScanError(t *testing.T) {
 		{`1.0x`, "invalid hexadecimal literal"},
 		{`0x0x`, "invalid hexadecimal literal"},
 		{`00x0x`, "invalid hexadecimal literal"},
-		{`9223372036854775808`, "value out of range"},
-		{`$9223372036854775808`, "value out of range"},
+		{`9223372036854775809`, "value out of range"},
```
Comment (tamird) [inline sql/parser/scan_test.go]: nit: *integer* value

**Comment by tamird** [comment]
LGTM

File: sql/parser/eval.go (hunk 1 above)
Comment (petermattis) [inline sql/parser/eval.go]: Well, `IntVal` is really storing an unsigned integer even though its underlying type is `int64`. I don't want to display the negative value. I think I can just use `%s` and rely on `IntVal.String`, though.

File: sql/parser/expr.go (hunk 2 above)
Comment (petermattis) [inline sql/parser/expr.go]: As I mentioned in the other comment, `IntVal` is storing an unsigned integer in an int64. I'll add an explanatory comment.

File: sql/parser/scan.go (hunk 3 above)
Comment (petermattis) [inline sql/parser/scan.go]: Yes.

File: sql/parser/scan_test.go (hunk 4 above)
Comment (petermattis) [inline sql/parser/scan_test.go]: Done.

**Pull request merged**